pkg/logger/json.go

```diff
@@ -0,0 +1,269 @@
+/*
+Copyright 2022 CodeNotary, Inc. All rights reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+	http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package logger
+
+import (
+	"bytes"
+	"encoding"
+	"encoding/json"
+	"fmt"
+	"io"
+	"os"
+	"runtime"
+	"sync"
+	"sync/atomic"
+	"time"
+)
+
+const (
+	// DefaultTimeFormat is the time format to use for JSON output
+	DefaultTimeFormat = "2006-01-02T15:04:05.000000Z07:00"
+
+	// errInvalidTypeMsg message implies an arg cannot be serialized to json
+	errInvalidTypeMsg = "cannot serialize arg(s) to json"
+
+	// defaultCallerOffset is the stack frame offset in the call stack for the caller
+	defaultCallerOffset = 4
+)
+
+var (
+	// DefaultOutput is used as the default log output.
+	DefaultOutput io.Writer = os.Stderr
+)
+
+var _ Logger = &JsonLogger{}
+
+// JsonLogger is a logger implementation for json logging.
+type JsonLogger struct {
+	name       string
+	timeFormat string
+
+	timeFnc TimeFunc
+
+	mutex  Locker
+	writer *writer
+
+	level *int32
+}
+
+// NewJsonLogger returns a json logger.
+func NewJsonLogger(opts *LoggerOptions) *JsonLogger {
+	if opts == nil {
+		opts = &LoggerOptions{}
+	}
+
+	output := opts.Output
+	if output == nil {
+		output = DefaultOutput
+	}
+
+	l := &JsonLogger{
+		name:       opts.Name,
+		timeFormat: DefaultTimeFormat,
+		timeFnc:    time.Now,
+		mutex:      new(sync.Mutex),
+		writer:     newWriter(output),
+		level:      new(int32),
+	}
+
+	if opts.TimeFnc != nil {
+		l.timeFnc = opts.TimeFnc
+	}
+	if opts.TimeFormat != "" {
+		l.timeFormat = opts.TimeFormat
+	}
+
+	atomic.StoreInt32(l.level, int32(opts.Level))
+
+	return l
+}
+
+func (l *JsonLogger) logWithFmt(name string, level LogLevel, msg string, args ...interface{}) {
+	msgStr := fmt.Sprintf(msg, args...)
+	l.log(name, level, msgStr)
+}
+
+// Log a message and a set of key/value pairs for a given level.
+func (l *JsonLogger) log(name string, level LogLevel, msg string, args ...interface{}) {
+	minLevel := LogLevel(atomic.LoadInt32(l.level))
+	if minLevel > level {
+		return
+	}
+
+	t := l.timeFnc()
+
+	l.mutex.Lock()
+	defer l.mutex.Unlock()
+
+	l.logJSON(t, name, level, msg, args...)
```

Comment:
@byo @jeroiraz I think eventually we can have a single logger (with json/plaintext as output), and can use the opts to configure which one to use. That way we wouldn't have to maintain many loggers. 

There can be another function, `logPlain` which handles memory/simple loggers. And file logger is just writing to another writer, and hence I've abstracted out in the struct to the output configurable.

This can be set for another story, but it will simplify the logger pkg

Reply:
https://github.com/codenotary/immudb-issues/issues/80